---
title: "Chapter X - R Notebook"
author: "Salom√© Carcy"
date: "`r format(Sys.Date(), '%d %B %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true

---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 5, fig.height = 5,
                      warning=FALSE, message=FALSE,
                      root.dir = "~/Projects/phd/")
```


# IMPORT

## Import librairies
```{r import-librairies}
library(ggplot2)
library(RColorBrewer)
library(cowplot)
library(tidyverse)
library(dplyr)
library(Seurat)
library(SCpubr)
library(patchwork)
library(DESeq2)
library(pheatmap)

source("~/Projects/phd/scripts/colors_universal.R")
```

## Import data
```{r import-data}
seur_integrated <- readRDS("./data/clean_data/seurat_human_integrated_object_23_12_01.rds")
DimPlot(seur_integrated, reduction="umap_integrated", group.by="clusters_integrated_data", cols=cols_integrated)
```




# FUNCTIONS
- `aggregate.Matrix` and `dMcast` functions taken from `Matrix.utils` package (https://rdrr.io/cran/Matrix.utils/src/R/Matrix.utils.R)
- `runDeseq` function written in-house, inspired from tutorial : https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html
```{r define-functions}
aggregate.Matrix<-function(x,groupings=NULL,form=NULL,fun='sum'){
  if(!is(x,'Matrix')){
    x<-Matrix(as.matrix(x),sparse=TRUE)
  }
  if(fun=='count'){
    x<-x!=0
  }
  groupings2<-groupings
  if(!is(groupings2,'data.frame')){
    groupings2<-as(groupings2,'data.frame')
  }
  groupings2<-data.frame(lapply(groupings2,as.factor))
  groupings2<-data.frame(interaction(groupings2,sep = '_'))
  colnames(groupings2)<-'A'
  if(is.null(form)){
    form<-as.formula('~0+.')
  }
  form<-as.formula(form)
  mapping<-dMcast(groupings2,form)
  colnames(mapping)<-substring(colnames(mapping),2)
  result<-t(mapping) %*% x
  if(fun=='mean'){
    result@x<-result@x/(aggregate.Matrix(x,groupings2,fun='count'))@x
  }
  attr(result,'crosswalk')<-grr::extract(groupings,match(rownames(result),groupings2$A))
  return(result)
}


dMcast<-function(data,formula,fun.aggregate='sum',value.var=NULL,as.factors=FALSE,factor.nas=TRUE,drop.unused.levels=TRUE){
  values<-1
  if(!is.null(value.var)){
    values<-data[,value.var]
    }
  alltms<-terms(formula,data=data)
  response<-rownames(attr(alltms,'factors'))[attr(alltms,'response')]
  tm<-attr(alltms,"term.labels")
  interactionsIndex<-grep(':',tm)
  interactions<-tm[interactionsIndex]
  simple<-setdiff(tm,interactions)
  i2<-strsplit(interactions,':')
  newterms<-unlist(lapply(i2,function (x) paste("paste(",paste(x,collapse=','),",","sep='_'",")")))
  newterms<-c(simple,newterms)
  newformula<-as.formula(paste('~0+',paste(newterms,collapse='+')))
  allvars<-all.vars(alltms)
  data<-data[,c(allvars),drop=FALSE]
  if(as.factors){
    data<-data.frame(lapply(data,as.factor))
    }
  characters<-unlist(lapply(data,is.character))
  data[,characters]<-lapply(data[,characters,drop=FALSE],as.factor)
  factors<-unlist(lapply(data,is.factor))
  #Prevents errors with 1 or fewer distinct levels
  data[,factors]<-lapply(data[,factors,drop=FALSE],function (x) 
  {
    if(factor.nas)
      if(any(is.na(x)))
      {
        levels(x)<-c(levels(x),'NA')
        x[is.na(x)]<-'NA'
      }
    if(drop.unused.levels)
        if(nlevels(x)!=length(na.omit(unique(x))))
          x<-factor(as.character(x))
    y<-contrasts(x,contrasts=FALSE,sparse=TRUE)
    attr(x,'contrasts')<-y
    return(x)
  })
  #Allows NAs to pass
  attr(data,'na.action')<-na.pass
  result<-Matrix::sparse.model.matrix(newformula,data,drop.unused.levels = FALSE,row.names=FALSE)
  brokenNames<-grep('paste(',colnames(result),fixed = TRUE)
  colnames(result)[brokenNames]<-lapply(colnames(result)[brokenNames],function (x) {
    x<-gsub('paste(',replacement='',x=x,fixed = TRUE) 
    x<-gsub(pattern=', ',replacement='_',x=x,fixed=TRUE) 
    x<-gsub(pattern='_sep = \"_\")',replacement='',x=x,fixed=TRUE)
    return(x)
  })

  result<-result*values
  if(isTRUE(response>0))
  {
    responses=all.vars(terms(as.formula(paste(response,'~0'))))
    result<-aggregate.Matrix(result,data[,responses,drop=FALSE],fun=fun.aggregate)
  }
  return(result)
}
```




# DE ANALYSIS BETWEEN THYMIC LINEAGES

## Prepare metadata and counts
```{r metadata-prep}
counts   <- seur_integrated@assays$RNA@counts
metadata <- seur_integrated@meta.data

# Subset metadata only to information we're interested in (tissue, cell identity, batch, and clusters)
metadata <- metadata[,c("tissue", "tcell_lineage", "batch_id", "clusters_integrated_data", "donor_id")]
colnames(metadata) <- c("tissue_id", "lineage_id", "batch_id", "cluster_id", "donor_id")
head(metadata) # sanity check

# Create a "groups" df that will (1) keep only cells of interest; (2) keep the columns of interest (that will define how we group counts)
groups <- metadata %>%
  rownames_to_column("cell") %>%
  # keep only cells of interest
  filter(tissue_id=="Thymus" & lineage_id %in% c("CD4", "iNKT", "CD8", "MAIT") & donor_id %in% 2:4) %>%
  # filter to clusters of interest
  filter(cluster_id %in% c(3,11,9,10)) %>%
  # keep only groups with at least 50 cells
  group_by(lineage_id, donor_id, cluster_id) %>%
  filter(n()>50) %>%
  ungroup() %>%
  # keep only columns of interest
  column_to_rownames("cell") %>%
  dplyr::select(lineage_id, donor_id, cluster_id) %>%
  # adapt a few variables
  mutate(cluster_id=as.character(cluster_id),
         donor_id=paste0("donor", donor_id))
head(groups) # sanity check
```

Prepare counts:
```{r counts-prep}
# AGGREGATE COUNTS
count.agg <- t(counts[, which(colnames(counts) %in% rownames(groups))]) # keep only cells of interest (defined in "groups") & put cells in rows (genes as columns)
nrow(groups) == nrow(count.agg) # verify nb of cells in "groups" (rows) is same nb of cells in "count.agg" (rows)
count.agg <- aggregate.Matrix(count.agg, groupings = groups, fun = "sum") # aggregate counts based on columns in "groups"

# Sanity checks
dim(count.agg)[2] == nrow(seur_integrated) # same nb of genes in seurat object & count.agg
count.agg[1:6, 1:6]
table(colSums(count.agg) == 0) # check if any gene has total count of 0

# Final counts for DESeq (groups as columns, genes as rows)
counts.deseq <- data.frame(t(count.agg))
counts.deseq <- counts.deseq[rowSums(counts.deseq)!=0,] # remove genes that have total count of 0


# PREPARE METADATA DF FOR DESEQ
# Have metadata df where rownames are the colnames of "counts.deseq" (the groups)
metadf.deseq <- groups %>%
  distinct() %>%
  unite(sample_id, remove=FALSE)
rownames(metadf.deseq) <- metadf.deseq$sample_id

# Check that the row names of metadf.deseq are the same as the column names of counts.deseq in order to use as input to DESeq2
nrow(metadf.deseq)==ncol(counts.deseq)
metadf.deseq <- metadf.deseq[match(colnames(counts.deseq), rownames(metadf.deseq)),] # reorder rows in metadf.deseq to match order of columns in counts.deseq
all(rownames(metadf.deseq) == colnames(counts.deseq))
```



## Run DE lineage by lineage
```{r analysis-2}
cellTypeSignature <- function(cell.1, norm="rlog", padj_max=0.05, log2FC_min=0.5, shrinkage=F, DEmethod="LRT"){
  
  # All cell lineages
  # allcells <- c("CD8", "CD4", "MAIT", "iNKT")
  # allothercells <- allcells[allcells != cell.1]
  if(cell.1%in%c("CD4", "iNKT")){allothercells=c("CD8", "MAIT")}
  else if(cell.1%in%c("CD8", "MAIT")){allothercells=c("CD4", "iNKT")}
  
  # Initialize
  genes.up <- list()
  
  # Loop
  for (cell.2 in allothercells){
    print(paste0(cell.1, "vs", cell.2))
    
    # 1. Create DDS object with only contrast of interest
    metadf.temp <- metadf.deseq %>%
      filter(lineage_id %in% c(cell.1, cell.2)) %>%
      mutate(lineage_id=factor(lineage_id, levels=unique(lineage_id)))
    counts.temp <- counts.deseq[,rownames(metadf.temp)]
    # table(colnames(counts.temp)==rownames(metadf.temp))
    dds.temp <- DESeqDataSetFromMatrix(counts.temp, colData = metadf.temp, design = ~ donor_id + lineage_id + cluster_id)
    
    # 2. Normalize, batch-correct counts and plot PCA
    # if(norm=="rlog"){rld.temp <- rlog(dds.temp, blind=TRUE)}
    # else if(norm=="vst"){rld.temp <- vst(dds.temp, blind=TRUE)}
    # counts_batchcorrect.temp <- limma::removeBatchEffect(x=assay(rld.temp),
    #                                                      batch=metadf.temp$batchdonor_id,
    #                                                      design=model.matrix(~ lineage_id + cluster_id, metadf.temp))
    # counts_pca.temp <- ddsPCA(counts.batchcorrect = counts_batchcorrect.temp, metadf=metadf.temp)
    # Plot PCA on batch-corrected counts
    # ggplot(counts_pca.temp, aes(x = PC1, y = PC2, color = lineage_id, shape=batch_id)) +
    #   geom_point(size = 4) +
    #   # xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) +
    #   # ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) +
    #   coord_fixed(expand=TRUE)+
    #   scale_color_manual(values=cols_lineages)
    # ggplot(counts_pca.temp, aes(x = PC1, y = PC2, color = cluster_id, shape=batch_id)) +
    #   geom_point(size = 4) +
    #   # xlab(paste0("PC1: ", round(percentVar[1] * 100), "% variance")) +
    #   # ylab(paste0("PC2: ", round(percentVar[2] * 100), "% variance")) +
    #   coord_fixed(expand=TRUE)+
    #   scale_color_manual(values=cols_integrated)
    
    # 3. Run DESeq2 (LRT test): parameters based on recommendations
    # https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#recommendations-for-single-cell-analysis
    dds.temp <- DESeq(dds.temp, test=DEmethod, reduced=~donor_id+cluster_id, useT=TRUE, minmu=1e-6, minReplicatesForReplace = Inf)
    # plotDispEsts(dds.temp) # Plot dispersion estimates
    contrastvec <- c("lineage_id", cell.1, cell.2)
    res.temp <- results(dds.temp, contrast = contrastvec, alpha = 0.05)
    if(shrinkage==T){
      print("Shrinking LFC")
      res.temp <- lfcShrink(dds.temp, type="ashr", contrast = contrastvec, res=res.temp)
    }
    # print(res.temp)
    genes.sig.temp <- res.temp %>% data.frame() %>% filter(padj<padj_max & log2FoldChange>log2FC_min)
    print(paste0("# upregulated genes:", nrow(genes.sig.temp)))
    
    # 4. Save list of DE genes
    genes.up[[paste0(cell.1, "vs", cell.2)]] <- rownames(genes.sig.temp)
  } # FOR LOOP
  
  return(genes.up)
}


# *************************************************
## 3.2. Get lineage-specific upregulated genes ####

# Get upregulated genes (note: shrinkage if Wald test, but LRT recommended for scRNAseq data)
cd4up  <- cellTypeSignature(cell.1="CD4",   padj_max=0.05, log2FC_min = 0.1, shrinkage=F)
cd8up  <- cellTypeSignature(cell.1="CD8",   padj_max=0.05, log2FC_min = 0.1, shrinkage=F)
maitup <- cellTypeSignature(cell.1="MAIT",  padj_max=0.05, log2FC_min = 0.1, shrinkage=F)
nktup  <- cellTypeSignature(cell.1="iNKT",  padj_max=0.05, log2FC_min = 0.1, shrinkage=F)
# gdtup  <- cellTypeSignature(cell.1="GD",    padj_max=0.05, log2FC_min = 0.1, shrinkage=F)

# Bind them
# genes.lineage <- unique(c(plyr::count(Reduce(c, cd4up))  %>% filter(freq>=3) %>% pull(x),
#                           plyr::count(Reduce(c, cd8up))  %>% filter(freq>=3) %>% pull(x),
#                           plyr::count(Reduce(c, maitup)) %>% filter(freq>=3) %>% pull(x),
#                           plyr::count(Reduce(c, nktup))  %>% filter(freq>=3) %>% pull(x)))
                          # plyr::count(Reduce(c, gdtup))  %>% filter(freq>=2) %>% pull(x)))

# Get intersect of DE genes
cd4_inkt_up <- intersect(
  plyr::count(Reduce(c, cd4up[c("CD4vsMAIT", "CD4vsCD8")]))  %>% filter(freq>=2) %>% pull(x),
  plyr::count(Reduce(c, nktup[c("iNKTvsMAIT", "iNKTvsCD8")]))  %>% filter(freq>=2) %>% pull(x)
)
cd8_mait_up <- intersect(
  plyr::count(Reduce(c, cd8up[c("CD8vsiNKT", "CD8vsCD4")]))  %>% filter(freq>=2) %>% pull(x),
  plyr::count(Reduce(c, maitup[c("MAITvsiNKT", "MAITvsCD4")])) %>% filter(freq>=2) %>% pull(x)
)
genes.lineage <- c(cd4_inkt_up, cd8_mait_up)

# genes.lineage <- unique(c(
#   plyr::count(Reduce(c, cd4up[c("CD4vsMAIT", "CD4vsCD8")]))  %>% filter(freq>=2) %>% pull(x),
#   plyr::count(Reduce(c, cd8up[c("CD8vsiNKT", "CD8vsCD4")]))  %>% filter(freq>=2) %>% pull(x),
#   plyr::count(Reduce(c, maitup[c("MAITvsiNKT", "MAITvsCD4")])) %>% filter(freq>=2) %>% pull(x),
#   plyr::count(Reduce(c, nktup[c("iNKTvsMAIT", "iNKTvsCD8")]))  %>% filter(freq>=2) %>% pull(x)
# ))
```


Get batch corrected counts with limma
```{r analysis-1}
# CREATE DESEQ2 OBJECT
dds <- DESeqDataSetFromMatrix(counts.deseq,
                              colData = metadf.deseq,
                              design = ~ donor_id + lineage_id + cluster_id)


# ****************************
## 2.2. PCA on the groups ####

# Transform counts for data visualization
rld <- rlog(dds, blind=TRUE)
# rld <- vst(dds)
# head(assay(rld))

# Plot PCA
# DESeq2::plotPCA(rld, intgroup = "sample_id")
# DESeq2::plotPCA(rld, intgroup = "batch_id")
# DESeq2::plotPCA(rld, intgroup = "lineage_id")

# Correct for batch effect
counts_batchcorrect <- limma::removeBatchEffect(
  x=assay(rld),
  batch=metadf.deseq$donor_id,
  design=model.matrix(~ lineage_id + cluster_id, metadf.deseq)
)
```


Plot heatmap
```{r pheatmap, fig.height=7, fig.width=8}
# Get corrected counts (that we batch corrected with limma earlier)
counts.correc.sig <- t(counts_batchcorrect[genes.lineage,]) # keep only the genes of interest
counts.correc.sig[,1:5]
dim(counts.correc.sig)

# Reorder by lineage and cluster (not by batch)
counts.correc.sig <- counts.correc.sig[metadf.deseq %>% arrange(lineage_id, as.numeric(cluster_id)) %>% pull(sample_id),]
cols_donorid <- c("lightgrey", "darkgrey", "#737373")
names(cols_donorid) <- unique(metadf.deseq$donor_id)

# Run pheatmap using the metadata data frame for the annotation
# pdf("./scripts-in-progress/human-thymus/HumanThymus_25_common_genes_CD4iNKT_CD8MAIT/plots/LRT_contrast1by1_clustersnaive_commonDEgenes_CD4iNKT_vs_CD8MAIT.pdf",
#     width=12, height=8)
pheatmap::pheatmap(
  t(counts.correc.sig),
  color = rev(colorRampPalette(brewer.pal(10, "RdBu"))(100)),
  scale = "row", # z-score
  clustering_method="ward.D2",
  cluster_rows = T,
  cluster_cols = T,
  border_color = NA,
  # Columns (cell groups)
  # gaps_col=c(10,18,27,32),
  show_colnames = T,
  fontsize_col = 6,
  annotation_col = metadf.deseq %>% dplyr::select(c(lineage_id, donor_id, cluster_id)),
  annotation_colors = list(
    lineage_id = cols_lineages,
    donor_id=cols_donorid,
    # batchdonor_id=c("lightgrey", "darkgrey"),
    cluster_id=cols_integrated[names(cols_integrated) %in% unique(metadf.deseq$cluster_id)]
  ),
  # Rows (genes)
  show_rownames=T,
  fontsize_row=6,
  # title
  main="donors 2:4, ncells>50, c3-9-10-11"
)
# dev.off()
```




# SESSION INFO
```{r}
sessionInfo()
```


