---
title: "Human T cell data - Correlation gene expression btw clusters of each T cell lineage"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE) 
```




## 1. IMPORT

### 1.1. Import librairies & data
```{r import-librairies, message=FALSE}
library(ggplot2)
library(cowplot)
library(tidyverse)
library(dplyr)
library(Seurat)
library(RColorBrewer)
library(harmony)
library(corrplot)
# library(ggalluvial)
# library(SingleCellExperiment)
```

```{r import-data}
# Import seurat object
# seur.human <- readRDS("~/Projects/HumanThymusProject/data/raw_data/human_data/filtered_seurat_harmony.11.22.22.RDS")
seur.filt <- readRDS("~/Projects/HumanThymusProject/data/CLUSTER/seurat-integration/input/seur_filt_22-12-19.rds") # removed low-quality cells

# Sanity check
print(seur.filt) # 80,614 cells

# Quick visualization
# DimPlot(seur.human, reduction="UMAP_50", group.by="new_clusters") +
#   scale_color_manual(values=colorRampPalette(brewer.pal(12,"Paired"))(18))
```




## 2. PREPROCESSING

### 2.1. Functions

```{r preprocessing1}
# Preprocessing
preprocess <- function(seurobj, ndim=20, res=0.8, celltype, approxPCA=TRUE){
  print(seurobj)
  cat("Cells:", unique(seurobj$group.ident), "\n")
  
  print("Normalize")
  seurobj <- NormalizeData(seurobj)
  print("HVG")
  seurobj <- FindVariableFeatures(seurobj, selection.method="vst", nfeatures=2000)
  print("Scale")
  seurobj <- ScaleData(seurobj)
  print("PCA")
  seurobj <- RunPCA(seurobj, npcs=50, verbose=FALSE, approx=approxPCA)
  print("Harmony")
  seurobj <- RunHarmony(seurobj, reduction = "pca", max.iter.harmony=30, group.by.vars = c("Method", "Batch"))
  # DimPlot(seurobj, reduction="pca", group.by="Batch") | DimPlot(seurobj, reduction="harmony", group.by="Batch")
  # DimPlot(seurobj, reduction="pca", group.by="Method") | DimPlot(seurobj, reduction="harmony", group.by="Method")
  print(ElbowPlot(seurobj, ndims=50, reduction="harmony"))
  print("UMAP")
  seurobj <- RunUMAP(seurobj, reduction = "harmony", dims = 1:ndim, n.neighbors=50, reduction.key = "UMAP50_")
  print("Cluster")
  seurobj <- FindNeighbors(seurobj, reduction = "harmony", dims = 1:ndim)
  seurobj <- FindClusters(seurobj, resolution = res, random.seed = 0)
  
  # Visualization sanity check
  print(DimPlot(seurobj, reduction = "umap", group.by="seurat_clusters", label = TRUE, repel = TRUE)+
    scale_color_manual(values=colorRampPalette(brewer.pal(12,"Paired"))(18)) +
    labs(title=celltype))
  
  return(seurobj)
}

# Sanity check
# test <- preprocess(seur.filt, res=1, celltype="all") # verified on 01.24
```


### 2.2. CD4 T cells
```{r preprocessing-cd4}
# Run preprocessing
unique(seur.filt$group.ident)
seur.cd4 <- preprocess(seurobj = subset(seur.filt, subset=group.ident=="CD4_Thymus"),  celltype = "CD4thy", ndim=10, res=0.2)

# Annotate
FindAllMarkers(seur.cd4, only.pos=TRUE, min.pct=0.25, logfc.threshold=0.5) %>%
  group_by(cluster) %>%
  slice_max(n=5, order_by=avg_log2FC)

seur.cd4@meta.data$cell_annot <- case_when(
  seur.cd4@meta.data$seurat_clusters == 0 ~ "thyCD4_ccr9",
  seur.cd4@meta.data$seurat_clusters == 1 ~ "thyCD4_ccr7",
  seur.cd4@meta.data$seurat_clusters == 2 ~ "thyCD4_ISP",
  seur.cd4@meta.data$seurat_clusters == 3 ~ "thyCD4_DPp",
  seur.cd4@meta.data$seurat_clusters == 4 ~ "thyCD4_Treg",
  seur.cd4@meta.data$seurat_clusters == 5 ~ "thyCD4_Tagonist",
  seur.cd4@meta.data$seurat_clusters == 6 ~ "thyCD4_DPq"
)

DimPlot(seur.cd4, group.by="cell_annot", label=T, repel=T) + theme(legend.position="none")
```


### 2.3. CD8 T cells
```{r preprocessing-cd8}
seur.cd8 <- preprocess(seurobj = subset(seur.filt, subset=group.ident=="CD8_Thymus"),  celltype = "CD8thy", ndim=10, res=0.2)

# Annotate
FindAllMarkers(seur.cd8, only.pos=TRUE, min.pct=0.25, logfc.threshold=0.5) %>%
  group_by(cluster) %>%
  slice_max(n=10, order_by=avg_log2FC)

seur.cd8@meta.data$cell_annot <- case_when(
  seur.cd8@meta.data$seurat_clusters == 0 ~ "thyCD8_ccr7",
  seur.cd8@meta.data$seurat_clusters == 1 ~ "thyCD8_ccr9", # ccr9?
  seur.cd8@meta.data$seurat_clusters == 2 ~ "thyCD8_cd8aa1",
  seur.cd8@meta.data$seurat_clusters == 3 ~ "thyCD8_idk",
  seur.cd8@meta.data$seurat_clusters == 4 ~ "thyCD8_cd8aa2",
  seur.cd8@meta.data$seurat_clusters == 5 ~ "thyCD8_DP"
)

DimPlot(seur.cd8, group.by="cell_annot", label=T, repel=T) + theme(legend.position="none")
```


### 2.4. MAIT cells
```{r preprocessing-mait}
# Preprocess
seur.mait <- preprocess(seurobj = subset(seur.filt, subset=group.ident=="MAIT_Thymus"), celltype = "MAITthy", ndim=10, res=0.2)
DimPlot(seur.mait)

# Annotate
FindAllMarkers(seur.mait, only.pos=TRUE, min.pct=0.25, logfc.threshold=0.5) %>%
  group_by(cluster) %>%
  slice_max(n=5, order_by=avg_log2FC)

seur.mait@meta.data$cell_annot <- case_when(
  seur.mait@meta.data$seurat_clusters == 0 ~ "thyMAIT_ccr7",
  seur.mait@meta.data$seurat_clusters == 1 ~ "thyMAIT_ccr9",
  seur.mait@meta.data$seurat_clusters == 2 ~ "thyMAIT_effector",
  seur.mait@meta.data$seurat_clusters == 3 ~ "thyMAIT_idk",
  seur.mait@meta.data$seurat_clusters == 4 ~ "thyMAIT_cd8aa",
  seur.mait@meta.data$seurat_clusters == 5 ~ "thyMAIT_DP",
  seur.mait@meta.data$seurat_clusters == 6 ~ "thyMAIT_IFNsig"
)

DimPlot(seur.mait, group.by="cell_annot", label=T) + theme(legend.position="none")
```


### 2.5. NKT cells
```{r preprocessing-nkt}
seur.nkt <- preprocess(seurobj = subset(seur.filt, subset=group.ident=="NKT_Thymus"),  celltype = "NKTthy", ndim=10, res=0.3)

# Annotate
FindAllMarkers(seur.nkt, only.pos=TRUE, min.pct=0.25, logfc.threshold=0.5) %>%
  group_by(cluster) %>%
  slice_max(n=5, order_by=avg_log2FC)

seur.nkt@meta.data$cell_annot <- case_when(
  seur.nkt@meta.data$seurat_clusters == 0 ~ "thyNKT_ccr7",
  seur.nkt@meta.data$seurat_clusters == 1 ~ "thyNKT_effFOSJUN",
  seur.nkt@meta.data$seurat_clusters == 2 ~ "thyNKT_ccr9",
  seur.nkt@meta.data$seurat_clusters == 3 ~ "thyNKT_cd8aa",
  seur.nkt@meta.data$seurat_clusters == 4 ~ "thyNKT_effector",
  seur.nkt@meta.data$seurat_clusters == 5 ~ "thyNKT_IFNsig"
)

DimPlot(seur.nkt, group.by="cell_annot", label=T, repel=T) + theme(legend.position="none")
```


### 2.6. GDT cells
```{r preprocessing-gdt}
seur.gd <- preprocess(seurobj = subset(seur.filt, subset=group.ident=="GD_Thymus"),   celltype = "GDthy", ndim=10, res=0.2)

# Annotate
FindAllMarkers(seur.gd, only.pos=TRUE, min.pct=0.25, logfc.threshold=0.5) %>%
  group_by(cluster) %>%
  slice_max(n=10, order_by=avg_log2FC)

seur.gd@meta.data$cell_annot <- case_when(
  seur.gd@meta.data$seurat_clusters == 0 ~ "thyGDT_IFNsig",
  seur.gd@meta.data$seurat_clusters == 1 ~ "thyGDT_ccr9",
  seur.gd@meta.data$seurat_clusters == 2 ~ "thyGDT_immat",
  seur.gd@meta.data$seurat_clusters == 3 ~ "thyGDT_effector",
  seur.gd@meta.data$seurat_clusters == 4 ~ "thyGDT_immat_cycl",
  seur.gd@meta.data$seurat_clusters == 5 ~ "thyGDT_DP"
)

DimPlot(seur.gd, group.by="cell_annot", label=T, repel=T) + theme(legend.position="none")
```


### 2.7. Merge back together

```{r preprocessing-all}
# Merge all cells back into one object
seur.thym <- merge(seur.cd4, y = c(seur.cd8, seur.mait, seur.nkt, seur.gd))
hvg <- unique(c(VariableFeatures(seur.cd4),
                VariableFeatures(seur.cd8),
                VariableFeatures(seur.mait),
                VariableFeatures(seur.nkt),
                VariableFeatures(seur.gd)))
length(hvg)
VariableFeatures(seur.thym) <- hvg

# Sanity checks
seur.thym # only 37,331 because no CD1a or CD1c
# seur.thym[["RNA"]]@counts
# seur.thym@meta.data
table(seur.thym@meta.data[,c("cell_annot", "group.ident")], useNA="ifany")
```




## 3. GENE CORRELATION

### 3.1. Average expression per gene & per cluster
```{r avgexp}
# Get average expression of each gene per cluster
Idents(seur.thym) <- "cell_annot"
avgexp <- AverageExpression(seur.thym, assays="RNA", slot="data")
avgexp <- as.data.frame(avgexp$RNA) # get it into a df
head(avgexp, 10)
dim(avgexp) # 32 clusters and 17,212 genes

# Keep only HVGs
avgexp <- avgexp[rownames(avgexp) %in% VariableFeatures(seur.thym),]
avgexp <- avgexp[rowSums (avgexp)!=0,] # remove genes not expressed

# Scale?
# avg = rowMeans(avgexp)
# avgexp = sweep(avgexp,1,avg,"/")
# rm(avg)
```


### 3.2. Compute correlation
```{r correlation-calc}
#7a:  Correlation
corr.method <- "spearman" # or pearson?
Corr.Coeff.Table = cor(avgexp, method=corr.method)

#7b:  Shuffle data
shuffled.cor.list = list()
pb   <- txtProgressBar(1, 100, style=3)
nPermutations <- 1000

# Calculate correlations on shuffled expression table
for (i in 1:nPermutations){
  # Shuffle gene exp by T lineage
  shuffled_cd4  = apply(avgexp[,str_detect(colnames(avgexp),"thyCD4_")],1,sample)
  shuffled_cd8  = apply(avgexp[,str_detect(colnames(avgexp),"thyCD8_")],1,sample)
  shuffled_mait = apply(avgexp[,str_detect(colnames(avgexp),"thyMAIT_")],1,sample)
  shuffled_nkt  = apply(avgexp[,str_detect(colnames(avgexp),"thyNKT_")],1,sample)
  shuffled_gdt  = apply(avgexp[,str_detect(colnames(avgexp),"thyGDT_")],1,sample)
  # Bind back together avg exp table (shuffled)
  shuffled = cbind(t(shuffled_cd4),t(shuffled_cd8),t(shuffled_mait),t(shuffled_nkt),t(shuffled_gdt))
  if(ncol(shuffled)!=ncol(avgexp)){ print("PROBLEM") }
  # Correlation on shuffled table
  shuffled.cor = cor(shuffled, method=corr.method)
  shuffled.cor.list[[i]] = shuffled.cor
  rm(list=c('shuffled_cd4','shuffled_cd8', 'shuffled_mait', 'shuffled_nkt', 'shuffled_gdt', 'shuffled', 'shuffled.cor'))
  if ((i %% 100) ==0){
    setTxtProgressBar(pb, (i*100)/nPermutations)
  }
}

  
# Make empty p.value and corr mean matrixes
p.value.table = matrix(ncol=ncol(avgexp), nrow = ncol(avgexp))
rownames(p.value.table) = colnames(avgexp)
colnames(p.value.table) = colnames(avgexp)

shuffled.mean.table = p.value.table

# Get the mean "random" correlation from all permutations, and pvalue comparing the "observed" correlation with the "random" one
a = combn(1:ncol(avgexp),2)
for (i in 1:ncol(a)){
  cor.scores = sapply(shuffled.cor.list,"[",a[1,i],a[2,i])
  # mean correlation score on random data
  shuffled.mean.table[a[1,i],a[2,i]] = mean(cor.scores)
  shuffled.mean.table[a[2,i],a[1,i]] = mean(cor.scores)
  # empirical p.val
  p.value = mean(abs(cor.scores)>=abs(Corr.Coeff.Table[a[1,i],a[2,i]]))
  p.value.table[a[1,i],a[2,i]] = p.value
  p.value.table[a[2,i],a[1,i]] = p.value
  rm(list=c('cor.scores','p.value'))
  setTxtProgressBar(pb, (i*100)/ncol(a))
}

# Sanity check
p.value.table[1:10,1:10]
shuffled.mean.table[1:10,1:10]
Corr.Coeff.Table[1:10,1:10]
# neg.log10.p = -log10(p.value.table)

avgexp
```


### 3.3. Plot correlation matrix
```{r plot-correlation, fig.height=10, fig.width=10}
# Prepare colors
col_scale <- colorRampPalette(c("white","darkred"))
col_text <- c(rep("#e41a1c", 7), # CD4
              rep("#377eb8", 6), # CD8
              rep("#4daf4a", 7), # MAIT
              rep("#984ea3", 6), # NKT
              rep("#ff7f00", 6)) # GDT
names(col_text) <- colnames(Corr.Coeff.Table)
# text_col <- text_col[corrMatOrder(Corr.Coeff.Table, order="hclust", hclust.method="ward.D")]

# Plot
jpeg("~/Projects/HumanThymusProject/data/human-thymus/HumanData_12_CorrelationGEbtwTlineages/corr_heatmap_ordered.jpeg", width=2200, height=2200, res=300)
corrplot(Corr.Coeff.Table,
         is.corr=F, # is input matrix a corr matrix?
         
         # Order as original
         order="original",
         # Order with HClustering 
         # order="hclust",
         # hclust.method = "ward.D",

         # Color scale
         method="color", # visualize corr as color
         # col=col_scale(200), # color scale
         col=kovesi.rainbow_bgyr_35_85_c73(200),
         col.lim=c(0.5,1),
         cl.align.text="l", # alignment text color legend
         diag=F, # no color in the diagonal
         
         # Significance
         p.mat=p.value.table,
         sig.level=0.05, # if pval>sig.level => corr coeff is insig
         # insig="pch", pch=19, pch.cex=0.25, pch.col="black", # add points to insig squares
         insig="blank", # make insignificant squares blank
         
         # Text & title
         main="Correlation on 5862 HVGs",
         tl.col=col_text, # text color
         tl.pos="lt", # position labels on left & top
         tl.cex=0.7, # size text label
         
         # Margins
         mar=c(3,1,5,1))
dev.off()

```



