---
title: "02_CorrelationComparison"
output: html_notebook
---

# IMPORT

```{r import-librairies}
library(ggplot2)
library(tidyverse)
library(biomaRt)
library(corrplot)
```

```{r import-data}
path <- "~/Projects/20220809_Thymic-iNKT-CrossSpecies/data/02_CorrelationComparion"
# Mouse data
DEgenes_ms.df <- readRDS(file.path(path, "ms_DEG.rds"))
ExpressionTable_ms <- readRDS(file.path(path, "ms_avgexp.rds"))
# Pig data
DEgenes_pig.df <- readRDS(file.path(path, "pig_DEG.rds"))
ExpressionTable_pig <- readRDS(file.path(path, "pig_avgexp.rds"))
```

# IDENTIFICATION OF ONE-TO-ONE ORTHOLOGS

```{r get-ENSEMBL-ID}
# It's better to input ENSEMBL ID in the biomart, so get a df of DEgenes with their ID & gene name

# Get mouse DEgenes EMSEMBL ID
DEgenes_ms.ref <- readr::read_tsv("~/Projects/20220809_Thymic-iNKT-CrossSpecies/data/raw_data/mouse_data/B6_1/features.tsv.gz", col_names = FALSE) %>%
  filter(X2 %in% DEgenes_ms.df$gene) %>%
  arrange(match(X2, DEgenes_ms.df$gene))
length(unique(DEgenes_ms.df$gene)) # 2819 unique DE genes
dim(DEgenes_ms.ref) # 2819 rows
length(unique(DEgenes_ms.ref$X2)) # 2818 unique DE genes (Atp5o.1 missing)
head(DEgenes_ms.ref)


# Get pig DEgenes EMSEMBL ID
DEgenes_pig.ref <- readr::read_tsv("~/Projects/20220809_Thymic-iNKT-CrossSpecies/data/raw_data/pig_data/GSE192520_RAW/iNKT/features.tsv.gz", col_names = FALSE) %>%
  filter(X2 %in% DEgenes_pig.df$gene) %>%
  arrange(match(X2, DEgenes_pig.df$gene))
length(unique(DEgenes_pig.df$gene)) # 2085 unique DE genes
dim(DEgenes_pig.ref) # 2085 rows
length(unique(DEgenes_pig.ref$X2)) # 2085 unique DE genes
head(DEgenes_pig.ref)
```

```{r get-ensembl-orthologs}
# Define marts to map mouse to human and ensembl to entrez IDs
# mart.hu <- useMart(biomart="ENSEMBL_MART_ENSEMBL",
#                    dataset="hsapiens_gene_ensembl")
mart.ms <- useMart(biomart="ENSEMBL_MART_ENSEMBL",
                       dataset="mmusculus_gene_ensembl")
mart.ss <- useMart(biomart="ENSEMBL_MART_ENSEMBL",
                       dataset="sscrofa_gene_ensembl")


# Get pig orthologs of mouse DEgenes
# listAttributes(mm_ensembl)[stringr::str_detect(listAttributes(mart.ms)$name, "sscrofa"),]
DEgenes_ms.ortholgs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name', 'sscrofa_homolog_ensembl_gene', 'sscrofa_homolog_associated_gene_name',
                                            'sscrofa_homolog_orthology_type', 'sscrofa_homolog_orthology_confidence'),
                             filters = 'ensembl_gene_id',
                             values = DEgenes_ms.ref$X1,
                             mart = mart.ms)
colnames(DEgenes_ms.ortholgs) <- c("ms_ensembl", "ms_gene", "pig_ensembl", "pig_gene", "sscrofa_homolog_orthology_type", "sscrofa_homolog_orthology_confidence")
# Remove genes that don't have a pig ortholog, and replace gene symbols with the ones from DEgenes_ms.ref (58 genes don't have the same symbol from biomart)
DEgenes_ms.ortholgs <- DEgenes_ms.ortholgs %>%
  filter(!is.na(sscrofa_homolog_orthology_confidence)) %>%
  mutate_all(na_if, "") %>%
  left_join(DEgenes_ms.ref[,1:2], by=c("ms_ensembl" = "X1")) %>%
  dplyr::select(-ms_gene) %>%
  rename(ms_gene=X2) %>%
  relocate(ms_gene, .after=ms_ensembl)
# Sanity checks
head(DEgenes_ms.ortholgs)
dim(DEgenes_ms.ortholgs) # 2781 rows
length(unique(DEgenes_ms.ortholgs$ms_ensembl)) # 2621 unique mouse genes (out of 2819)
# length(unique(DEgenes_ms.ortholgs$ms_gene)) # 2621 unique mouse genes
DEgenes_ms.ref[DEgenes_ms.ref$X1 %in% setdiff(DEgenes_ms.ref$X1, DEgenes_ms.ortholgs$ms_ensembl),] # 198 genes didn't have any known pig ortholog
table(DEgenes_ms.ortholgs$sscrofa_homolog_orthology_confidence) # 499 pig genes have 0 orthology confidence
table(DEgenes_ms.ortholgs$ms_ensembl %in% DEgenes_ms.ref$X1) # in the orthologs table all mouse genes have the same name in the DEgenes table
table(DEgenes_ms.ortholgs$ms_gene %in% DEgenes_ms.ref$X2)




# Get mouse orthologs of pig DEgenes
# listAttributes(ss_ensembl)[stringr::str_detect(listAttributes(mart.ss)$name, "mmusc"),]
DEgenes_pig.ortholgs <- getBM(attributes = c('ensembl_gene_id', 'external_gene_name', 'mmusculus_homolog_ensembl_gene', 'mmusculus_homolog_associated_gene_name',
                                             'mmusculus_homolog_orthology_type', 'mmusculus_homolog_orthology_confidence'),
                              filters = 'ensembl_gene_id',
                              values = DEgenes_pig.ref$X1,
                              mart = mart.ss)
colnames(DEgenes_pig.ortholgs) <- c("pig_ensembl", "pig_gene", "ms_ensembl", "ms_gene", "mmusculus_homolog_orthology_type", "mmusculus_homolog_orthology_confidence")
# Remove genes that don't have a mouse ortholog, and replace pig gene symbols with the ones from DEgenes_pig.ref (some genes don't have the same symbol from biomart)
DEgenes_pig.ortholgs <- DEgenes_pig.ortholgs %>%
  filter(!is.na(mmusculus_homolog_orthology_confidence)) %>%
  mutate_all(na_if,"") %>%
  left_join(DEgenes_pig.ref[,1:2], by=c("pig_ensembl" = "X1")) %>%
  dplyr::select(-pig_gene) %>%
  rename(pig_gene=X2) %>%
  relocate(pig_gene, .after=pig_ensembl)
# Sanity checks
head(DEgenes_pig.ortholgs)
dim(DEgenes_pig.ortholgs) # 2082 rows
length(unique(DEgenes_pig.ortholgs$pig_ensembl)) # 1922 unique pig genes (out of 2085)
# length(unique(DEgenes_pig.ortholgs$pig_gene)) # 1922 pig gene symbols
table(DEgenes_pig.ortholgs$mmusculus_homolog_orthology_confidence) # 398 mouse genes have 0 orthology confidence
table(DEgenes_pig.ortholgs$pig_ensembl %in% DEgenes_pig.ref$X1) # all pig genes have the same name in the orthologs table and the DEgenes table
table(DEgenes_pig.ortholgs$pig_gene %in% DEgenes_pig.ref$X2)
```




# COMPARATIVE ANALYSIS

Step 1: Take intersect, union, species1 or species2 of DEgenes for analysis

```{r intersect}
# # Define the DEgenes (let's choose the ms_ensembl as reference)
# DEgenesSpecies1 = DEgenes_ms.ortholgs$ms_ensembl
# # nDESp1 = length(unique(DEgenesSpecies1))
# DEgenesSpecies2 = DEgenes_pig.ortholgs$ms_ensembl
# # nDESp2 = length(unique(DEgenesSpecies2))
# 
# 
# # Get the union or intersection between genes from Species1/2 (or keep only genes from Species 1 or Species 2)
# #genes.use = c('intersect','union','species1','species2')
# if (genes.use=='intersect') DEgenes = intersect(DEgenesSpecies1,DEgenesSpecies2)
# if (genes.use=='union')     DEgenes = union(DEgenesSpecies1,DEgenesSpecies2)
# if (genes.use=='species1') DEgenes = DEgenesSpecies1
# if (genes.use=='species2') DEgenes = DEgenesSpecies2
# 
# 
# # Verify how many genes we got
# length(unique(DEgenes)) # 3786 genes

# Identify overlapping DEgenes in mouse & pig (take mouse as reference)
DEgenes <- intersect(DEgenes_ms.ortholgs$ms_ensembl, DEgenes_pig.ortholgs$ms_ensembl)
length(unique(DEgenes)) # 1088
table(is.na(DEgenes))


# Verify if any of these DEgenes ms_ensembl have several pig orthologs
DEgenes_pig.ortholgs %>%
  filter(ms_ensembl %in% DEgenes) %>%
  group_by(ms_ensembl) %>%
  filter(n_distinct(pig_ensembl)>1)
# ENSMUSG00000068252 ms gene (Apol7b) => unsure if pig ENSSSCG00000000148 or ENSSSCG00000039706
# ENSMUSG00000071716 ms gene (Apol7e) => unsure if pig ENSSSCG00000000148 or ENSSSCG00000039706
# ENSMUSG00000019961 ms gene (Tmpo)   => unsure if pig ENSSSCG00000000887 or ENSSSCG00000038487 (TMPO)
DEgenes_ms.ortholgs %>%
  filter(pig_ensembl %in% c("ENSSSCG00000000148", "ENSSSCG00000039706", "ENSSSCG00000000887", "ENSSSCG00000038487"))
# ENSMUSG00000019961 ms gene (Tmpo) is the pig ENSSSCG00000038487 (TMPO) with an orthology confidence of 1

# Let's remove the other genes as unsure what their pig orthologs are
DEgenes <- DEgenes[! DEgenes %in% c("ENSMUSG00000068252", "ENSMUSG00000071716")]
length(DEgenes) # 1086

# Verify if any of these DEgenes ms_ensemble have the same pig gene ortholog
DEgenes_pig.ortholgs %>%
  filter(ms_ensembl %in% DEgenes) %>%
  group_by(pig_ensembl) %>%
  filter(n_distinct(ms_ensembl)>1)
# ENSSSCG00000004024 pig gene => unsure if mouse Rnaset2b (ENSMUSG00000094724) or Rnaset2a (ENSMUSG00000095687)
# ENSSSCG00000006418 pig gene => unsure if mouse Ifi203 (ENSMUSG00000039997) or Ifi209 (ENSMUSG00000043263)
# ENSSSCG00000016475 pig gene => unsure if mouse Trbc1 (ENSMUSG00000076490) or Trbc2 (ENSMUSG00000076498)
# TAF9 pig gene => unsure if mouse Ak6 (ENSMUSG00000078941) or TAF9 (ENSMUSG00000052293)
# BCL2A1 pig gene => unsure if mouse Bcl2a1b (ENSMUSG00000089929) or Bcl2a1d (ENSMUSG00000099974)
dbl_genes <- c("ENSMUSG00000094724", "ENSMUSG00000095687", "ENSMUSG00000039997", "ENSMUSG00000043263", "ENSMUSG00000076490",
               "ENSMUSG00000076498", "ENSMUSG00000078941", "ENSMUSG00000052293", "ENSMUSG00000089929", "ENSMUSG00000099974")
DEgenes_ms.ortholgs %>%
  filter(ms_ensembl %in% dbl_genes) %>%
  arrange(pig_ensembl)
# all of these mouse genes have some overlapping pig orthologs... so we'll remove them from analysis
DEgenes <- DEgenes[! DEgenes %in% dbl_genes]
length(DEgenes) # 1076


# Get the equivalent ms/pig gene names that are DE expressed and we keep for downstream comparative analysis
DEgenes_ms.list  <- unique(DEgenes_ms.ortholgs[DEgenes_ms.ortholgs$ms_ensembl %in% DEgenes,"ms_gene"])
DEgenes_pig.list <- unique(DEgenes_pig.ortholgs[DEgenes_pig.ortholgs$ms_ensembl %in% DEgenes,"pig_gene"])
DEgenes_pig.list <- DEgenes_pig.list[! DEgenes_pig.list == "ENSSSCG00000000887"] # we keep only TMPO and not ENSSSCG00000000887 (line 153)
length(DEgenes_ms.list) # 1076 msDE genes
length(DEgenes_pig.list) # 1076 pigDE genes
```




Step 2: Prune Expression Tables & Remove rows with no expression

```{r prune-expression-tables}
# Sanity check
# table(DEgenes_ms.list %in% rownames(ExpressionTable_ms))
# table(DEgenes_pig.list %in% rownames(ExpressionTable_pig))

# Prune expression tables
Sp1 = ExpressionTable_ms[rownames(ExpressionTable_ms) %in% DEgenes_ms.list,]
Sp1 = Sp1[rowSums (Sp1)!=0,]
Sp2 = ExpressionTable_pig[rownames(ExpressionTable_pig) %in% DEgenes_pig.list,]
Sp2 = Sp2[rowSums (Sp2)!=0,]
```




Step 3: Replace pig gene names with mouse gene names

```{r replace-pig-gene-names}
# Add new column with pig gene names
Sp2[,ncol(Sp2)+1] = rownames(Sp2)
colnames(Sp2)[ncol(Sp2)] = "pig_gene"
# Add mouse ortholog genes, keeping only the ones from the DEgenes list
Sp2 <- merge(Sp2, DEgenes_pig.ortholgs[,c("pig_gene", "ms_ensembl")], by="pig_gene", all.y = F) %>%
  distinct() %>%
  filter(ms_ensembl %in% DEgenes) %>%
  # get ms_gene from DEgenes_ms.ref table (otherwise it's the biomart ms_gene)
  left_join(DEgenes_ms.ortholgs[, c("ms_ensembl", "ms_gene")], by="ms_ensembl") %>%
  distinct()
rownames(Sp2) = Sp2$ms_gene
# Remove gene columns (keep only expression matrix)
Sp2 = Sp2[,!colnames(Sp2) %in% c("pig_gene", "ms_ensembl", "ms_gene")]
```




Step 4: Scale Expression Tables by gene average

```{r scale-expression-tables}
avg = rowMeans(Sp1)
Sp1 = sweep(Sp1,1,avg,"/")
rm(avg)

avg = rowMeans(Sp2)
Sp2 = sweep(Sp2,1,avg,"/")
rm(avg)
```



Step 5: Merge Expression tables
```{r merge-expression-tables}
# Sanity check
setdiff(rownames(Sp1), rownames(Sp2))
setdiff(rownames(Sp2), rownames(Sp1))

# Merge Expression Tables
geTable = merge(Sp1,Sp2, by='row.names', all=F)
rownames(geTable) = geTable$Row.names
geTable = geTable[,2:ncol(geTable)]
```



Step 6: Correlation

```{r correlation}
#7a:  Correlation
corr.method <- "spearman" # or pearson?
Corr.Coeff.Table = cor(geTable,method=corr.method)


#7b:  Shuffle data
shuffled.cor.list = list()
pb   <- txtProgressBar(1, 100, style=3)
nPermutations <- 1000

for (i in 1:nPermutations){
  shuffled = apply(geTable[,1:ncol(Sp1)],1,sample)
  shuffled2 = apply(geTable[,(ncol(Sp1)+1):ncol(geTable)],1,sample)
  shuffled = cbind(t(shuffled),t(shuffled2))
  shuffled.cor = cor(shuffled,method=corr.method)
  shuffled.cor.list[[i]] = shuffled.cor
  rm(list=c('shuffled','shuffled2','shuffled.cor'))
  if ((i %% 100) ==0){
    setTxtProgressBar(pb, (i*100)/nPermutations)
  }
}
  
p.value.table = matrix(ncol=ncol(geTable), nrow = ncol(geTable))
rownames(p.value.table) = colnames(geTable)
colnames(p.value.table) = colnames(geTable)
  
shuffled.mean.table = matrix(ncol=ncol(geTable), nrow = ncol(geTable))
rownames(shuffled.mean.table) = colnames(geTable)
colnames(shuffled.mean.table) = colnames(geTable)
  
a = combn(1:ncol(geTable),2)
for (i in 1:ncol(a)){
  cor.scores = sapply(shuffled.cor.list,"[",a[1,i],a[2,i])
  shuffled.mean.table[a[1,i],a[2,i]] = mean(cor.scores)
  shuffled.mean.table[a[2,i],a[1,i]] = mean(cor.scores)
  p.value = mean(abs(cor.scores)>=abs(Corr.Coeff.Table[a[1,i],a[2,i]]))
  p.value.table[a[1,i],a[2,i]] = p.value
  p.value.table[a[2,i],a[1,i]] = p.value
  rm(list=c('cor.scores','p.value'))
  setTxtProgressBar(pb, (i*100)/ncol(a))
}

neg.log10.p = -log10(p.value.table)
```




Step 7: Overlap in markers
For all pairs of cell-types generate list of genes that are at least 1.5x avg in both cells

```{r overlap-in-markers}
#from above a = combn(1:ncol(geTable),2)
marker.overlap.list = list()
for (i in 1:ncol(a)){
  datasubset = cbind(geTable[,a[1,i]],geTable[,a[2,i]])
  markers = rownames(geTable[datasubset[,1]>1.5 & datasubset[,2]>1.5,])
  marker.overlap.list[[i]] = markers
  names(marker.overlap.list)[i] = paste(colnames(geTable)[a[1,i]], colnames(geTable)[a[2,i]],sep='_')
  rm(list=c('datasubset','markers'))
}

# list.to.return = list(Corr.Coeff.Table,shuffled.mean.table,p.value.table,neg.log10.p,
#                       DEgenes,rownames(geTable),length(DEgenes),length(rownames(geTable)),
#                       nDESp1,nDESp2,geTable,marker.overlap.list)
# names(list.to.return) = c('corr.coeff','shuffled_correlation_score_means','p.value','negative_log10_p.value',
#                           'DEgenes_intersect','DEgenes_in_analysis','nDEgenes_intersect','nDEgenes_in_analysis',
#                           'nDEgenes_Sp1','nDEgenes_Sp2','scaled_table','overlapping_markers')
```




Step 8: Plot
```{r plot-correlation}
# Not sure what this does...
Method1="intersect"
# comp.intersect <- SpPermute(ExpressionTableSpecies1, species1=Species1, DEgenesSpecies1, ExpressionTableSpecies2, species2=Species2, DEgenesSpecies2, nPermutations=1000, genes.use= Method1,corr.method="spearman")
comp_table.intersect <- t(Corr.Coeff.Table[1:ncol(ExpressionTable_ms), (ncol(ExpressionTable_ms)+1):nrow(Corr.Coeff.Table)])
p_table.intersect <- t(p.value.table[1:ncol(ExpressionTable_ms), (ncol(ExpressionTable_ms)+1):nrow(Corr.Coeff.Table)])
p_table.intersect <- (-p_table.intersect)
	

# Plot
col1 <- colorRampPalette(c("darkblue", "white","darkred"))
corrplot(comp_table.intersect, order="original",tl.pos="lt", method="color", tl.col="black",
         cl.lim=c(min(comp_table.intersect),max(comp_table.intersect)),
         is.corr=F, tl.cex=0.7, sig.level=(-0.05), insig="pch", pch=19, pch.cex=0.25, pch.col="black",
         p.mat=p_table.intersect, col=col1(200), main= paste(Method1,",",length(rownames(geTable)), "genes", sep=" "),mar=c(3,1,5,1),cl.align.text="l")

```


